### PR TITLE
Fix #356 by moving a where clause

### LINKED ISSUE #356: Final metapuzzle doesn't appear in puzzle list
When I look at the list of puzzles I've authored (https://puzzlehunt.azurewebsites.net/1/author/Puzzles), my metameta doesn't appear.  It used to appear on this page, so this is a regression.

### PR DESCRIPTION
**Pull request by tabascq**
By moving the filetype where-clause before the join instead of after,  we get the right results and what I hope is equivalent-or-better perf.

### CODE REVIEW

**Comment by morganbr** [comment]
@tabascq , can you please post the queries from your debugger output that this produced?

**Comment by tabascq** [comment]
I have absolutely no idea how to do that. Give me detailed instructions and I will do my best to follow them.

**Comment by tabascq** [comment]
Oh wait, you mean the queries that are *already there* in my output. Sure, one sec :-)

**Comment by morganbr** [comment]
Debug one of the pages and paste your debugger's output pane here

**Comment by tabascq** [comment]
I think it's this one:

SELECT [p].[ID], [p].[CustomURL], [p].[Description], [p].[EventID], [p].[Group], [p].[HintCoinsForSolve], [p].[HintsAreCumulative], [p].[IsCheatCode], [p].[IsFinalPuzzle], [p].[IsGloballyVisiblePrerequisite], [p].[IsMetaPuzzle], [p].[IsPuzzle], [p].[MaxAnnotationKey], [p].[MinPrerequisiteCount], [p].[MinutesOfEventLockout], [p].[MinutesToAutomaticallySolve], [p].[Name], [p].[OrderInGroup], [p].[PuzzleVersion], [p].[SolveValue], [p].[SupportEmailAlias], [p].[Token], [t].[ID], [t].[EventID], [t].[FileType], [t].[PuzzleID], [t].[ShortName], [t].[UrlString]
FROM [Puzzles] AS [p]
LEFT JOIN (
    SELECT [j].[ID], [j].[EventID], [j].[FileType], [j].[PuzzleID], [j].[ShortName], [j].[UrlString]
    FROM [ContentFiles] AS [j]
    WHERE [j].[FileType] = 0
) AS [t] ON [p].[ID] = (
    SELECT TOP(1) [subQuery0].[ID]
    FROM [Puzzles] AS [subQuery0]
    WHERE [subQuery0].[ID] = [t].[PuzzleID]
)

File: ServerCore/Pages/Puzzles/Index.cshtml.cs [hunk 1]
```diff
@@ -40,9 +40,8 @@ public async Task<IActionResult> OnGetAsync()
             }
 
             Puzzles = await (from Puzzle p in query
-                             join ContentFile joinFile in _context.ContentFiles on p equals joinFile.Puzzle into fileJoin
+                             join ContentFile joinFile in _context.ContentFiles.Where((j) => j.FileType == ContentFileType.Puzzle) on p equals joinFile.Puzzle into fileJoin
```
Comment (morganbr) [inline ServerCore/Pages/Puzzles/Index.cshtml.cs]: Mixing the two kinds of LINQ syntax is a bit weird. I'm not certain of the right syntax, so you can leave this as-is if you need to. My guess on the right syntax would be:

```
from Puzzle p in query
let puzzleFiles = from content in _context.ContentFiles
where content.FileType == ContentFileType.Puzzle
join ContentFile joinFile in puzzleFiles on p equals joinFile.Puzzle into fileJoin
from ContentFile file in fileJoin.DefaultIfEmpty()
orderby p.Group, p.OrderInGroup, p.Name
select new PuzzleView { Puzzle = p, Content = file }).ToListAsync();
```

**Comment by tabascq** [comment]
Jay, I also find joins to be completely inscrutable as well and cannot write one to save my life. But I think I know just enough to make this work. A join takes two inputs, and I prefiltered one of those inputs (the one from ContentFiles) so it doesn't ever see materials, etc.

**Comment by morganbr** [comment]
Just realized the new query is probably not filtering the content files by event. That produces the right output, but may be slow when we have multiple events. You might need to filter the ContentFile's event Id

**Comment by tabascq** [comment]
Morgan, I can barely read the LINQ syntax you use and do not feel remotely comfortable writing it. That may well be it but I'm afraid to try it because I certainly can't apply it to the other one.

**Comment by jaylorch** [comment]
One thing that puzzles me about the join is why DefaultIfEmpty doesn't need a parameter saying what the default *is* if empty.  It seems like it magically knows that the default if empty should be an element with all the puzzle fields as you want them, instead of (as I would think) a default join element that has all the puzzle fields null.

**Comment by morganbr** [comment]
@jaylorch it's using the default .NET value for a class, which is null.

**Comment by tabascq** [comment]
I added the event filter (using the only syntax I can write)

**Comment by morganbr** [comment]
Can we see the new debug output?

**Comment by tabascq** [comment]
SELECT [p].[ID], [p].[CustomURL], [p].[Description], [p].[EventID], [p].[Group], [p].[HintCoinsForSolve], [p].[HintsAreCumulative], [p].[IsCheatCode], [p].[IsFinalPuzzle], [p].[IsGloballyVisiblePrerequisite], [p].[IsMetaPuzzle], [p].[IsPuzzle], [p].[MaxAnnotationKey], [p].[MinPrerequisiteCount], [p].[MinutesOfEventLockout], [p].[MinutesToAutomaticallySolve], [p].[Name], [p].[OrderInGroup], [p].[PuzzleVersion], [p].[SolveValue], [p].[SupportEmailAlias], [p].[Token], [t].[ID], [t].[EventID], [t].[FileType], [t].[PuzzleID], [t].[ShortName], [t].[UrlString]
FROM [Puzzles] AS [p]
LEFT JOIN (
    SELECT [j].[ID], [j].[EventID], [j].[FileType], [j].[PuzzleID], [j].[ShortName], [j].[UrlString]
    FROM [ContentFiles] AS [j]
    WHERE ([j].[EventID] = @__Event_1_ID) AND ([j].[FileType] = 0)
) AS [t] ON [p].[ID] = (
    SELECT TOP(1) [subQuery0].[ID]
    FROM [Puzzles] AS [subQuery0]
    WHERE [subQuery0].[ID] = [t].[PuzzleID]
)
WHERE [p].[EventID] = @__Event_0_ID
ORDER BY [p].[Group], [p].[OrderInGroup], [p].[Name], [p].[ID]

**Comment by morganbr** [comment]
K. It looks like it might be doing some extra subqueries, but it's probably the best we can do. Ship it.

**Comment by jaylorch** [comment]
For anyone else who was curious, I found the pattern of using DefaultIfEmpty for a left outer join documented here:  https://docs.microsoft.com/en-us/dotnet/csharp/linq/perform-left-outer-joins

**Pull request merged**